### ISSUE TITLE
TiledMap performance

### ISSUE THREAD

**Issue by MateuszGawel** [libgdx/libgdx#2779]
Hello, I spent a lot of time on optimizing my game and I have one point I want to say/discuss as an potential issue possible to be upgraded.

It's not a breaking news when I'll say that every additional tileLayer influences the performance pretty highly. In my game (15x10) tiles on the screen, sized (64x64 px) with culling (default implementation) rendering tiled map takes 15% of whole game where's a lot of box2d, actors, animations etc. I have only 3 tile layers but for sure I'm gonna need 2 more. With 9 layers it's more than 1/4 of total game. The worst thing is that it's even not rendered because most of them are mainly empty for testing purposes.

I had a feeling that it can be upgraded and it seems that this is the bottleneck:
	<code>
public Cell getCell (int x, int y) {
		if (x < 0 || x >= width) return null;
		if (y < 0 || y >= height) return null;
		return cells[x][y];
	}
</code>
in TiledMapTileLayer. This method is being called for every tile visible on the screen for all 3 layers.so It gives 15x10x3=450 calls per frame in my case. The cell array is pretty big because the level is big. 

I think that making this array cells[y][x] should make a small improvement because most screens are wide and it's better to have 10 arrays of 15 elements than 15 arrays of 10 elements. Or it will be good to have an option to choose if our game is vertical or horizontal. 

Also it is worth to consider if we read the array in the most efficent way (in the order how it's stored in memory) and if the cells are stored in the same order how we want to read them. This is more complicated case because it may be stored in different ways on different devices (iOS i.e) but in Java there are arrays of arrays where elements of one array are continous.

Finally, reading from such big 2d array may be a problem itself. In my case there's a 500x40 tiled map. It's not a big deal but for 995 calls i got 1142 CPU time (DDMS) so it gives ~1ms per call. I use Motorola Moto G with quad core processor 1,2GHz so imagine how it can influence older devices. I think that some kind of partitioning may help if previous ones can't.

What do you think guys about it?

PS getCell is about 60% of whole tiledmap render(). About 30% is the method itself. I didn't measure times for every lines but there is a lot of other work implemented there so I don't think that there's any bottleneck. Worth to check though.

**Comment by MobiDevelop** [comment]
I've used the current implementation with up to 10 layers and not seen any significant slowdown, even on older single core devices. We can certainly take a look, but I'd be surprised if that is an actual meaningful bottleneck.

**Comment by MateuszGawel** [comment]
It's not so significant. Going from 3 to 9 layers gives about 5%-10% better performance. It may be imperceptible but when you're optimizing the game every 5% matters. Also the tiledMap level is important here.

I can send you some screenshots or profiled data if you need proof or some data to base on.

**Comment by NathanSweet** [comment]
Seems odd for that to be the bottleneck, but you might try `cells[y + x *
w]`.

**Comment by MateuszGawel** [comment]
Well, it's pretty hard to override LibGDX classes builded with gradle. I can try this one but I will have to create many classes because TiledMapLayer is used in other LibGDX classes which are used in other etc.

**Comment by NathanSweet** [comment]
You would change the libgdx source.

**Comment by MateuszGawel** [comment]
can you tell me what's the simplest way to do that when i have my project builded with gradle and libGDX is already builded. I don't want to set up everything from the beginning.

**Comment by NathanSweet** [comment]
You want to run from source without setting anything up?

**Comment by MateuszGawel** [comment]
I don't want to but if there's no clever way to do that with gradle I think I will have to.

**Comment by Tom-Ski** [comment]
https://github.com/libgdx/libgdx/wiki/Running-demos-%26-tests

**Comment by MobiDevelop** [comment]
Anyway, I benchmarked each of the following storage / access approaches...

* `[width][height]` / `[x][y]`
* `[height][width]` / `[y][x]`
* `[height * width]` / `[y * width + x]`

I set TiledMapBench to create a 20 layer, 500 x 40 map, approximately 15 x 10 viewport. As before, getCell only accounted for ~0.1% of CPU time for each approach - with virtually no difference between them.

Test device: Nexus 5, Android 5.0.1.

**Comment by MateuszGawel** [comment]
I benchmarked it too and got no improvement at all after implementing it on 1D array. :(
But look on that screenshot from DDMS. 19104ms is total time when getCell take 3172ms (2167 calls). It's 16% CPU time, how is it possible that you have 0.1%. I've tested it on 4 different phones and have the same problem.

![getcell](https://cloud.githubusercontent.com/assets/2331520/5971311/ba5373a8-a842-11e4-9c80-ca4531269c46.png)

Am I doing somethign wrong? I can attach my tiled map and code when I render it (95% of it is untouched LibGDX code) if you are kind enough to help.
